### PR TITLE
Implement least squares channel sorting

### PR DESCRIPTION
**Pull request by UberPyro**

### CODE REVIEW

File: bot.py [hunk 1]
```diff
@@ -40,6 +42,41 @@ def get_archive_category(guild):
     return discord.utils.get(guild.categories, name="Archive")
 
 
+def score(arr): 
+    return sum(e**2 for e in arr)
+
+
+def idx_div(arr, idxs): 
+    return [sum(arr[i:j]) for i, j in zip((0,) + idxs, idxs + (len(arr),))]
```
Comment (the-moonwitch) [inline bot.py:50]: what is `arr`? what is `idxs`? what does this function do?

File: bot.py [hunk 2]
```diff
@@ -40,6 +42,41 @@ def get_archive_category(guild):
...
+
+# `arr` is an array of sizes
+# `m` is the number of groups
+# returns indices that divides `arr` up
+# into groups of roughly the same size
+def balance_categories(arr, m): 
+    m -= 1
+    global best
+    global top_score
```
Comment (the-moonwitch) [inline bot.py:59]: not merging anything with `global`

File: bot.py [hunk 3]
```diff
@@ -40,6 +42,41 @@ def get_archive_category(guild):
...
+
+def idx_div(arr, idxs): 
+    return [sum(arr[i:j]) for i, j in zip((0,) + idxs, idxs + (len(arr),))]
+
+# `arr` is an array of sizes
+# `m` is the number of groups
+# returns indices that divides `arr` up
+# into groups of roughly the same size
+def balance_categories(arr, m): 
+    m -= 1
```
Comment (the-moonwitch) [inline bot.py:57]: what's the point of passing `m` if you want `m-1`?

File: bot.py [hunk 4]
```diff
@@ -40,6 +42,41 @@ def get_archive_category(guild):
...
+    global best
+    global top_score
+    best = (0,) * m
+    top_score = maxsize
+    arr_len = len(arr)
+
+    def bal_cat_rec(m, i, cur_indices):
```
Comment (the-moonwitch) [inline bot.py:64]: what's `m`? what's `i`? what's `cur_indices`? what does this function return?

File: bot.py [hunk 5]
```diff
@@ -70,27 +106,18 @@ async def sort(ctx: discord.ext.commands.Context):
...
-            else:
-                while channels[end_idx - 1].name[0].upper() == end_letter:
-                    end_idx -= 1
-                end_letter = channels[end_idx - 1].name[0].upper()
+    # balanced categorizer
+    letters = [ch.name[0].upper() for ch in channels]
+    alpha_counter = sorted(Counter("".join(letters)).items())
+    counts = list(map(lambda x: x[1], alpha_counter))
+    idxs = balance_categories(counts, len(categories))
+    for k, (i, j) in enumerate(zip([0] + idxs, idxs + [len(counts)])):
```
Comment (the-moonwitch) [inline bot.py:114]: this loop is way too complex and should probably be simplified into something understandable

File: bot.py (hunk 1 above)
Comment (UberPyro) [reply bot.py:50]: The function takes in an arbitrary array and a list of indices, slices the array between all the indices and the sums the slices.

File: bot.py (hunk 2 above)
Comment (UberPyro) [reply bot.py:59]: Ok, I'll try to get rid of it.

File: bot.py (hunk 3 above)
Comment (UberPyro) [reply bot.py:57]: `m` is the desired number of categories, `m-1` is the number of divisors you need in the array to get m categories. `m` to me makes more sense.

File: bot.py (hunk 4 above)
Comment (UberPyro) [reply bot.py:64]: The function takes in the number of divisors left to be placed, the current index of the input array, and the current list of indices as candidates for divisors (which may be incomplete at this point).

File: bot.py (hunk 5 above)
Comment (UberPyro) [reply bot.py:114]: Ok

**Comment by the-moonwitch** [comment]
Rejecting cause I wrote a slightly shorter implementation in https://github.com/slavfox/ChannelSorter2/commit/d36ea6e76604017872a4d0b080551ac175dd191f. Have credited you in the README, thanks for the PR!

**Pull request closed without merging**